OOps/goto_ops.c

```diff
@@ -41,8 +41,10 @@ int32_t kgoto(CSOUND *csound, GOTO *p)
   CS_PDS = p->lblblk->prvp;
   /* VL 16.2.23 fix for UDOs where the
      label gets confused */
-  if(CS_PDS->insdshead == NULL) 
```

Comment:
Note to self: review history of this to see what this originally fixed and whether it is still a problem.

Reply:
Marking as not resolved as I haven't looked into it yet.

Reply:
Looks like this code was added in this commit: https://github.com/csound/csound/commit/91b1ed0d4006ec15b79798f25c37ee027e1574e4 

Which mentions it was done to fix loops in UDOs. I found the mailing list thread that this was about:

https://listserv.heanet.ie/cgi-bin/wa?A2=ind2302&L=CSOUND&P=70026

I think this needs a little more investigation to address both the loops bug as well as deinit problem.

Reply:
can we add an issue to do this and move on?

Reply:
Ok I got to the bottom of this after barking at wrong trees:

- INDS and OPDS are often cast one to the other (BAD IDEA!)
- When I modified OPDS to include the deinit chain, a misalignment was caused
- in this case when we reassigned CS_PDS->insdshead, we were writing to an INSDS object and overwriting it caused the nxtoff to be rewritten 
- that made timexpire() go into an infinite loop.

Still figuring out the best solution. I fixed it by adding a padding member, but that's a horrible hack.

Reply:
There are two solutions:

1) Don't allow writing to CS_PDS->insdshead: that was always overwriting something, which previously was symptomless, but still wrong. One wonders how it could have ever worked, but the insdshead pointer seems to have been carried through correctly.

2) Add an insdshead to INSDS at the same byte offset to OPDS: it's a bit of a hack but should ensure that it can be used to carry insdshead.

Now the REAL solution is to disentangle INSDS and OPDS. This is a major piece of work that requires a dedicated PR. 

Incidentally, I think the original cause of the bug is due to INSDS being read as OPDS in the case of UDOs. I have not found the exact point, but I'd say we should find it in one of the useropcd() functions.

Reply:
I looked and didn't see casts between the two, mostly casting of void* to OPDS*. Casting between the two definitely seems wrong. Could you point me to somewhere that CS_PDS->insdshead is being written to? I'm not keen on #2 as I think it'd make this issue even more difficult to disentangle in the future. 

At this point, if the code is restored, does that still cause deact to go into a loop? Seems we also need to add a test CSD for both the loops in UDOs as well as deact. 

And how would you like to proceed? I won't mind if we file an issue and move forward, as I want to get to the PRs that restore the opcodes that were removed (e.g., "active", which my live code system uses and I need to better test things).

Reply:
There's the casting of INSDS to OPDS in various places, for example in insert.c:2728,  `instance()`, we have

```
prvids = prvpds = prvpdd = (OPDS*) ip;
```
In the case of UDOs, `p->lblblk->prvp`  is NULL, but it is an `INDS *` not a `OPDS *`. `CS_PDS` is `p->h.insdshead->pds`, which is a `OPDS *`. That assignment is OK if we just use it for jumping, but in the case of the fix I applied, it assumes we have an `OPDS *` when we have an `INSDS *` instead. 

It was a fluke that it worked, because it carried the correct data for jumping and it overwrite something that did not cause a problem. When I added the deinit members in the OPDS, that caused the overwriting to be in a place where it mattered (for deact)

So, we can remove the wrong fix and try a better one for the case of loops in UDOs, or we can embrace the OPDS/INSDS collusion by making sure it works in this case.

This is a case of the old code coming back to bite us, I think. Not sure yet how hard it is to disentangle it.

Reply:
I think at this point we need to get through a lot of PRs and better to take on the regression. I've filed an issue here:

https://github.com/csound/csound/issues/1879

and assigned it to the two of us since we've been looking at it. 

Going to approve and merge this PR and move on to the next one.